### PR TITLE
Initial implementation of pipeline visual editor

### PR DESCRIPTION
**Pull request by Pavinati**
Drag and drop blocks and connect blocks together to build a pipeline

image

Signed-off-by: Mattia Pavinati <mattia.pavinati@ispirata.com>

### CODE REVIEW

File: src/react/components/VisualFlowEditor.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  engine.getNodeFactories().registerFactory(nodeFactory);
+  engine.setModel(model);
+
+  return engine;
+}
+
+const VisualFlowEditor = ({className = "", blocks, model, nodeFactory, onNodeSettingsClick}) => {
+  const [engine, setEngine] = useState(getEngine(model, nodeFactory));
+
+  const addBlock = useCallback((name, position) => {
```
Comment (davidebriani) [inline src/react/components/VisualFlowEditor.js:72]: Can we have `model` and `nodeFactory` generated on component's mount instead on delegating the work to the parent component and relying on props?

File: src/react/NewPipelinePage.js [hunk 2]
```diff
@@ -40,6 +89,18 @@ export default ({ astarte, history }) => {
   });
   const formAlerts = useAlerts();
 
+  useEffect(() => {
+    astarte
+      .getBlocks()
+      .then(astarteBlocks => {
+        setBlocks(astarteBlocks);
+        nodeFactory.updateDefinitions(astarteBlocks);
```
[the remark this answers is not in the document]
Comment (Pavinati) [reply src/react/NewPipelinePage.js:97]: NodeFactory is created at the beginning and is required for the editor to launch. It's node/block knowledge is updated as soon as we get more info from Astarte

File: src/react/NewPipelinePage.js [hunk 3]
```diff
@@ -16,22 +16,71 @@
...
+      </JsonSchemaForm>
+    </Modal.Body>
+  </Modal>
+);
 
 export default ({ astarte, history }) => {
   const [isCreatingPipeline, setIsCreatingPipeline] = useState(false);
+  const [blocks, setBlocks] = useState([]);
+  const [editorModel, setEditorModel] = useState(getNewModel());
+  const [nodeFactory, setNodeFactory] = useState(astarteNodeFactory());
```
[the remark this answers is not in the document]
Comment (Pavinati) [reply src/react/NewPipelinePage.js:82]: This is required to preserve the status when removing the editor from the view. Unless we implement status serialization/deserialization, we can't have it inside the editor or we'll lose it

File: src/react/NewPipelinePage.js [hunk 4]
```diff
@@ -16,22 +16,71 @@
...
+  astarteNodeFactory,
+  getNewModel,
+  VisualFlowEditor,
+} from "./components/VisualFlowEditor.js";
 import SingleCardPage from "./ui/SingleCardPage.js";
 
 const ajv = new Ajv({schemaId: 'id'});
-ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
+const metaSchemaDraft04 = require("ajv/lib/refs/json-schema-draft-04.json");
+ajv.addMetaSchema(metaSchemaDraft04);
```
[the remark this answers is not in the document]
Comment (Pavinati) [reply src/react/NewPipelinePage.js:47]: it is required for schema checking

File: src/react/components/VisualFlowEditor.jsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    prevBlock = nextBlock;
+
+    loopCounter += 1;
+    if (loopCounter > 50) {
+      throw new Error('Pipeline too long');
+    }
+  } while (nextBlock.blockType !== 'consumer');
+
+  return chain.map((block) => block.toScript()).join('\n| ');
+}
```
Comment (davidebriani) [inline src/react/components/VisualFlowEditor.jsx:161]: Would it make sense to move this method directly into `models/NativeBlockModel.js` ?

Comment (Pavinati) [reply src/react/components/VisualFlowEditor.jsx:161]: This has to do with the pipeline as a whole, not with the single block. I'd leave it in the editor itself, as it is the one that defines and controls the connection between nodes

**Review by davidebriani — APPROVED** [review]

**Pull request merged**